### PR TITLE
Feat/basemodel utils

### PR DESCRIPTION
**Pull request by xValentim**
Oi professor, 

passando o código que comentei na aula. Separei num arquivo ``utils.py`` pra ficar mais visível o que fiz. 

main changes:

- [x] BaseModel class pra encapsular os métodos fit, compile e evaluate. 
- [x] Plot class pra encapsular plots a respeito do history (metrics and train/val losses)
- [x] validation_split como parametro do .fit()
- [ ] TODO: callback de Early stopping
- [x] Metadados no treino loss, val_loss e metrcis (keras-like)

### CODE REVIEW

**Comment by xValentim** [comment]
Subi um notebook mostrando teste e deixei o log das células em ``notebooks/testing_utils.ipynb`` com o link [aqui](https://github.com/xValentim/my_nlp_library/blob/main/notebooks/testing_utils.ipynb)

```python

sample_size = 10_000
embedding_dim = 32

dataset_train, dataset_test, tokenizer, _ = nlp.get_imdb_dataset(sample_size=sample_size)
vocab_size = tokenizer.vocab_size

model = nlp.ExperimentModel(vocab_size=vocab_size, 
                            embedding_dim=embedding_dim, 
                            output_dim=1)

model.compile(optimizer=torch.optim.Adam(model.parameters(), lr=1e-3),
              device="cuda",
              metrics=["accuracy", 
                       "precision", 
                       "recall", 
                       "f1"])

history = model.fit(dataset_train, 
                   n_epochs=25,
                   validation_split=0.2)

print("Training finished")
```

**Comment by xValentim** [comment]
O plot fica legalzinho, usando 

```python 
model.plot_history(dark_mode=True)
```

image

**Comment by xValentim** [comment]
Oi professor,

Aproveitando o espaço aqui pra mandar o comentario que fiz na aula:

[Proximal Policy Optimization](https://arxiv.org/pdf/1707.06347)
[Proximal Policy Optimization (PPO): The Key to LLM Alignment (ainda não li esse mas parece ser uma boa explicação](https://cameronrwolfe.substack.com/p/proximal-policy-optimization-ppo)
